File: src/CCVTAC.Console/ExternalTools/Runner.cs
using System.Diagnostics;

namespace CCVTAC.Console.ExternalTools;

internal static class Runner
{
    private const int AuthenticSuccessExitCode = 0;

    private static bool IsSuccessExitCode(HashSet<int> otherSuccessExitCodes, int exitCode) =>
        otherSuccessExitCodes.Append(AuthenticSuccessExitCode).Contains(exitCode);

    /// <summary>
    /// Calls an external application.
    /// </summary>
    /// <param name="settings"></param>
    /// <param name="otherSuccessExitCodes">Additional exit codes, other than 0, that can be treated as non-failures.</param>
    /// <param name="printer"></param>
    /// <returns>A `Result` containing the exit code, if successful, or else an error message.</returns>
    internal static Result<(int SuccessExitCode, string Warnings)> Run(
        ToolSettings settings,
        HashSet<int> otherSuccessExitCodes,
        Printer printer
    )
    {
        Watch watch = new();

        printer.Info($"Running {settings.CommandWithArgs}...");

        var splitCommandWithArgs = settings.CommandWithArgs.Split([' '], 2);

        ProcessStartInfo processStartInfo = new()
        {
            FileName = splitCommandWithArgs[0],
            Arguments = splitCommandWithArgs.Length > 1 ? splitCommandWithArgs[1] : string.Empty,
            UseShellExecute = false,
            RedirectStandardOutput = false,
            RedirectStandardError = true,
            CreateNoWindow = true,
            WorkingDirectory = settings.WorkingDirectory,
        };

        using Process? process = Process.Start(processStartInfo);

        if (process is null)
        {
            return Result.Fail($"Could not locate {splitCommandWithArgs[0]}.");
        }

        string errors = process.StandardError.ReadToEnd(); // Must precede `WaitForExit()`
        process.WaitForExit();
        printer.Info($"{splitCommandWithArgs[0]} finished in {watch.ElapsedFriendly}.");

        var trimmedErrors = errors.TrimTerminalLineBreak();
        return IsSuccessExitCode(otherSuccessExitCodes, process.ExitCode)
            ? Result.Ok((process.ExitCode, trimmedErrors)) // Errors will be considered warnings.
            : Result.Fail(
                $"{splitCommandWithArgs[0]} exited with code {process.ExitCode}: {trimmedErrors}."
            );
    }
}

File: src/CCVTAC.FSharp/ExternalTools/Runner.fs
namespace CCVTAC.Console.ExternalTools

open CCVTAC.Console
open Startwatch.Library
open System
open System.Diagnostics

module Runner =

    type ToolResult = { ExitCode: int; Error: string option }

    [<Literal>]
    let private authenticSuccessExitCode = 0

    let private isSuccessExitCode (otherSuccessExitCodes: int list) (exitCode: int) =
        List.contains exitCode (authenticSuccessExitCode :: otherSuccessExitCodes)

    /// Calls an external application.
    /// <param name="settings">Tool settings for execution</param>
    /// <param name="otherSuccessExitCodes">Additional exit codes, other than 0, that can be treated as non-failures</param>
    /// <param name="printer">Printer for logging</param>
    /// <returns>A Result instance containing the exit code and any warnings or else an error message.</returns>
    let internal run toolSettings (otherSuccessExitCodes: int list) (printer: Printer)
        : Result<ToolResult, string> =

        let watch = Watch()
        printer.Info $"Running {toolSettings.CommandWithArgs}..."

        let splitCommandWithArgs = toolSettings.CommandWithArgs.Split([|' '|], 2)

        let processStartInfo = ProcessStartInfo splitCommandWithArgs[0]
        processStartInfo.Arguments <- if splitCommandWithArgs.Length > 1
                                      then splitCommandWithArgs[1]
                                      else String.Empty
        processStartInfo.UseShellExecute <- false
        processStartInfo.RedirectStandardOutput <- false
        processStartInfo.RedirectStandardError <- true
        processStartInfo.CreateNoWindow <- true
        processStartInfo.WorkingDirectory <- toolSettings.WorkingDirectory

        match Process.Start processStartInfo with
        | null ->
            Error $"Could not locate {splitCommandWithArgs[0]}."
        | process' ->
            let error = process'.StandardError.ReadToEnd()

            process'.WaitForExit()
            printer.Info $"{splitCommandWithArgs[0]} finished in {watch.ElapsedFriendly}."

            let trimmedErrors = if String.hasText error
                                then Some (String.trimTerminalLineBreak error)
                                else None

            if isSuccessExitCode otherSuccessExitCodes process'.ExitCode
            then Ok { ExitCode = process'.ExitCode; Error = trimmedErrors }
            else Error $"{splitCommandWithArgs[0]} exited with code {process'.ExitCode}: {trimmedErrors}."
